### PR TITLE
Constellation informs script about documents becoming inactive, active or fully active.

### LINKED ISSUE #14876: Determine if a Document is active
(After  #14559 merges)
We currently can only check if a document is `fully active`; however, there are places in the spec that require us to check if the document is just `active` (example: https://html.spec.whatwg.org/multipage/webappapis.html#dom-document-write step 3 whatwg/html#2216). We need to implement a flag to keep track of this.

cc @asajeffrey

### PR DESCRIPTION
**Pull request by asajeffrey**
This PR replaces the current freeze/thaw mechanism by messages from the constellation to script informing it about when documents become inactive, active or fully active.

This means we can now implement |Document::is_active()| which is used in |document.write|.

This PR also changes how timers work: previously they were initialized running, and were then frozen/thawed. This means there was a transitory period when timers were running even though the document was not fully active. With this PR, timers are initially suspended, and are only resumed when the document is made fully active.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #14876 
- [X] These changes do not require tests because it's an interal refactoring.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14971)

### CODE REVIEW

**Comment by asajeffrey** [comment]
cc @jdm @cbrewster

**Comment by asajeffrey** [comment]
Hmm, there are wpt failures for this, I've still not quite got timers sorted out. I'll let you know once I've got something that passes wpt locally.

**Comment by asajeffrey** [comment]
I rewrote this PR to explicitly initialize the document activity, which means we can go back to timers initially running. This fixes the wpt failures locally.

This PR is now ready for review.

**Comment by asajeffrey** [comment]
Oh good, an intermittent:
```
./mach test-wpt --release --repeat-until-unexpected ~/_mozilla/mozilla/websocket_connection_fail.html
...
Running 1 tests in web-platform-tests

Ran 185 tests finished in 1.0 seconds.
  • 185 ran as expected. 0 tests skipped.

Running 1 tests in web-platform-tests
...
  ▶ Unexpected subtest result in /_mozilla/mozilla/websocket_connection_fail.html:
  │ TIMEOUT [expected PASS] Untitled
  └   → Test timed out
```

**Comment by asajeffrey** [comment]
That intermittent is reproducable on master: #14985.

**Comment by asajeffrey** [comment]
Rebased.

**Comment by jdm** [comment]
@Ms2ger Review ping.

**Comment by Ms2ger** [comment]
Can we add a test for the `document.write()` change?

r? @cbrewster

**Comment by asajeffrey** [comment]
@Ms2ger OK, I'll give it a shot. The test might fail at the moment, because writing into iframe content documents isn't fully implemented IIRC. @nox?

**Comment by nox** [comment]
@asajeffrey I have no idea what you mean by "fully implemented". Why would iframe content documents be different than any other document? As long is it's same origin I don't see what's not implemented.

**Comment by asajeffrey** [comment]
@nox: I pushed a test to this PR (https://github.com/servo/servo/pull/14971/commits/90fb8e335c52b523f7cab953c0f2ab297cf6802b) which will show you what I mean. It passes in FF, but fails in servo with the error 'The object is in an invalid state'.

**Review by cbrewster — COMMENTED** [review]
These changes look good! Just a question about when we should be updating the activity on pipelines.

Is the test working, or is it still a WIP?

File: components/constellation/constellation.rs [hunk 1]
```diff
@@ -1405,10 +1405,6 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
 
             let window_size = old_pipeline.and_then(|old_pipeline| old_pipeline.size);
 
-            if let Some(old_pipeline) = old_pipeline {
```
Comment (asajeffrey) [inline components/constellation/constellation.rs]: @jdm: thoughts? Should we suspend timers when a document is navigated from, or when it becomes inactive? (I think the latter, but you may have opinions.)

Comment (jdm) [inline components/constellation/constellation.rs]: There is explicit text about clearing timers when unloading a document: https://html.spec.whatwg.org/multipage/browsers.html#unloading-document-cleanup-steps step 3.2, invoked from step 10 of https://html.spec.whatwg.org/multipage/browsers.html#unload-a-document.

Comment (jdm) [inline components/constellation/constellation.rs]: However, for the cases in which the document is salvageable, I believe this is handled in the spec by ignoring timer tasks for documents which are not fully active in step 1 of https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-processing-model . Therefore, when the document becomes inactive would be equivalent.

Comment (asajeffrey) [inline components/constellation/constellation.rs]: OK, so this change is spec-compatible then. Thanks!

**Comment by asajeffrey** [comment]
Rebased.

**Comment by cbrewster** [comment]
```
  ▶ Unexpected subtest result in /html/dom/dynamic-markup-insertion/document-write/write-active-document.html:
  │ FAIL [expected PASS] document.write only writes to active documents
  └   → The object is in an invalid state.
```

**Comment by asajeffrey** [comment]
Oops, forgot to check in the `.ini` file.

**Pull request merged**